File: test/fileTypes.test.js
import { test, describe } from 'node:test';
import assert from 'node:assert';
import {
  getFileExtension,
  isJpegFile,
  isTiffFile,
  isDngFile,
  isRawFile,
  isSupportedFile,
  getFileType
} from '../src/exif/fileTypes.js';

describe('File Types', () => {
  test('getFileExtension should extract extension correctly', () => {
    assert.strictEqual(getFileExtension('/path/to/file.jpg'), 'jpg');
    assert.strictEqual(getFileExtension('/path/to/file.JPEG'), 'jpeg');
    assert.strictEqual(getFileExtension('/path/to/file.tiff'), 'tiff');
    assert.strictEqual(getFileExtension('/path/to/file'), '');
    assert.strictEqual(getFileExtension('/path/to/.hidden'), '');
  });

  test('isJpegFile should detect JPEG files', () => {
    assert.strictEqual(isJpegFile('/path/to/image.jpg'), true);
    assert.strictEqual(isJpegFile('/path/to/image.jpeg'), true);
    assert.strictEqual(isJpegFile('/path/to/image.JPG'), true);
    assert.strictEqual(isJpegFile('/path/to/image.png'), false);
    assert.strictEqual(isJpegFile('/path/to/image.tiff'), false);
  });

  test('isTiffFile should detect TIFF files', () => {
    assert.strictEqual(isTiffFile('/path/to/image.tiff'), true);
    assert.strictEqual(isTiffFile('/path/to/image.tif'), true);
    assert.strictEqual(isTiffFile('/path/to/image.TIFF'), true);
    assert.strictEqual(isTiffFile('/path/to/image.jpg'), false);
  });

  test('isDngFile should detect DNG files', () => {
    assert.strictEqual(isDngFile('/path/to/image.dng'), true);
    assert.strictEqual(isDngFile('/path/to/image.DNG'), true);
    assert.strictEqual(isDngFile('/path/to/image.jpg'), false);
  });

  test('isRawFile should detect RAW files', () => {
    assert.strictEqual(isRawFile('/path/to/image.dng'), true);
    assert.strictEqual(isRawFile('/path/to/image.cr2'), true);
    assert.strictEqual(isRawFile('/path/to/image.nef'), true);
    assert.strictEqual(isRawFile('/path/to/image.arw'), true);
    assert.strictEqual(isRawFile('/path/to/image.jpg'), false);
  });

  test('isSupportedFile should detect all supported formats', () => {
    assert.strictEqual(isSupportedFile('/path/to/image.jpg'), true);
    assert.strictEqual(isSupportedFile('/path/to/image.tiff'), true);
    assert.strictEqual(isSupportedFile('/path/to/image.dng'), true);
    assert.strictEqual(isSupportedFile('/path/to/image.cr2'), true);
    assert.strictEqual(isSupportedFile('/path/to/image.png'), false);
    assert.strictEqual(isSupportedFile('/path/to/image.gif'), false);
  });

  test('getFileType should return correct type', () => {
    assert.strictEqual(getFileType('/path/to/image.jpg'), 'jpeg');
    assert.strictEqual(getFileType('/path/to/image.tiff'), 'tiff');
    assert.strictEqual(getFileType('/path/to/image.dng'), 'dng');
    assert.strictEqual(getFileType('/path/to/image.cr2'), 'raw');
    assert.strictEqual(getFileType('/path/to/image.png'), 'unknown');
  });
});

File: tests/file_types_test.rs
use ifex::exif::file_types::FileType;
use std::path::Path;

#[test]
fn test_file_type_from_path_jpeg() {
  assert_eq!(
    FileType::from_path(Path::new("test.jpg")),
    Some(FileType::Jpeg)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.jpeg")),
    Some(FileType::Jpeg)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.JPG")),
    Some(FileType::Jpeg)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.JPEG")),
    Some(FileType::Jpeg)
  );
}

#[test]
fn test_file_type_from_path_tiff() {
  assert_eq!(
    FileType::from_path(Path::new("test.tif")),
    Some(FileType::Tiff)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.tiff")),
    Some(FileType::Tiff)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.TIF")),
    Some(FileType::Tiff)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.TIFF")),
    Some(FileType::Tiff)
  );
}

#[test]
fn test_file_type_from_path_dng() {
  assert_eq!(
    FileType::from_path(Path::new("test.dng")),
    Some(FileType::Dng)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.DNG")),
    Some(FileType::Dng)
  );
}

#[test]
fn test_file_type_from_path_raw() {
  assert_eq!(
    FileType::from_path(Path::new("test.cr2")),
    Some(FileType::Raw)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.nef")),
    Some(FileType::Raw)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.arw")),
    Some(FileType::Raw)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.orf")),
    Some(FileType::Raw)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.rw2")),
    Some(FileType::Raw)
  );
  assert_eq!(
    FileType::from_path(Path::new("test.raf")),
    Some(FileType::Raw)
  );
}

#[test]
fn test_file_type_from_path_unsupported() {
  assert_eq!(FileType::from_path(Path::new("test.png")), None);
  assert_eq!(FileType::from_path(Path::new("test.txt")), None);
  assert_eq!(FileType::from_path(Path::new("test")), None);
}

#[test]
fn test_supports_direct_exif() {
  assert!(FileType::Jpeg.supports_direct_exif());
  assert!(FileType::Tiff.supports_direct_exif());
  assert!(!FileType::Dng.supports_direct_exif());
  assert!(!FileType::Raw.supports_direct_exif());
}

#[test]
fn test_supports_dng_processing() {
  assert!(!FileType::Jpeg.supports_dng_processing());
  assert!(!FileType::Tiff.supports_dng_processing());
  assert!(FileType::Dng.supports_dng_processing());
  assert!(!FileType::Raw.supports_dng_processing());
}

#[test]
fn test_requires_sidecar() {
  assert!(!FileType::Jpeg.requires_sidecar());
  assert!(!FileType::Tiff.requires_sidecar());
  assert!(!FileType::Dng.requires_sidecar());
  assert!(FileType::Raw.requires_sidecar());
}

#[test]
fn test_as_str() {
  assert_eq!(FileType::Jpeg.as_str(), "jpeg");
  assert_eq!(FileType::Tiff.as_str(), "tiff");
  assert_eq!(FileType::Dng.as_str(), "dng");
  assert_eq!(FileType::Raw.as_str(), "raw");
}
